src/storage/remote/read.c

```diff
@@ -120,33 +117,13 @@ storageReadRemote(THIS_VOID, Buffer *buffer, bool block)
             {
                 MEM_CONTEXT_TEMP_BEGIN()
                 {
-                    PackRead *const read = protocolClientDataGet(this->client);
-                    pckReadNext(read);
-
-                    // If binary then read the next block
-                    if (pckReadType(read) == pckTypeBin)
-                    {
-                        MEM_CONTEXT_OBJ_BEGIN(this)
-                        {
-                            this->block = pckReadBinP(read);
-                            this->remaining = bufUsed(this->block);
-                        }
-                        MEM_CONTEXT_OBJ_END();
-                    }
-                    // Else read is complete and get the filter list
-                    else
-                    {
-                        bufFree(this->block);
-
-                        ioFilterGroupResultAllSet(ioReadFilterGroup(storageReadIo(this->read)), pckReadPackP(read));
-                        this->eof = true;
+                    if (!protocolClientSessionQueued(this->session))
+                        protocolClientSessionRequestAsyncP(this->session);
 
-                        protocolClientDataEndGet(this->client);
-                    }
+                    storageReadRemoteInternal(this, protocolClientSessionResponse(this->session));
```

Comment:
It waits until the async read completes. So on the first pass through we issue the async read and sit and wait for it because there is nothing else useful for us to do. However, after that first read we issue another async request and then go to work on the first response. By the time we are done with that we hope the second response will be waiting for us to read.

Reply:
OK, I'll update that then. Just leave this comment open for now. I don't like to push commits while review is happening because it can move thing around. This one wouldn't, but...